test/tests/functional/pbs_hook_execjob_prologue.py

```diff
@@ -0,0 +1,216 @@
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+# A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
+# details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# The PBS Pro software is licensed under the terms of the GNU Affero General
+# Public License agreement ("AGPL"), except where a separate commercial license
+# agreement for PBS Pro version 14 or later has been executed in writing with
+# Altair.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software - under a
+# commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+from tests.functional import *
+
+
+class TestPbsExecutePrologue(TestFunctional):
+    """
+    This tests the feature in PBS that allows execjob_prologue hook to
+    execute on all sister moms all the time, and not just when first
+    task is spawned on the node.
+
+    PRE: Have a cluster of PBS with 3 mom hosts.
+    """
+    def setUp(self):
+        if len(self.moms) != 3:
+            self.skip_test(reason="need 3 mom hosts: -p moms=<m1>:<m2>:<m3>")
+            return
+
+        TestFunctional.setUp(self)
+
+        self.momA = self.moms.values()[0]
+        self.momB = self.moms.values()[1]
+        self.momC = self.moms.values()[2]
+
+        self.hostA = self.momA.shortname
+        self.hostB = self.momB.shortname
+        self.hostC = self.momC.shortname
+
+        self.server.expect(VNODE, {'state=free': 3}, op=GE, max_attempts=10,
+                           interval=2)
+
+    def test_prologue_execute_on_all_moms(self):
+        """
+        Test to make sure execjob_prologue always get
+        executed on al sister moms when mother superior
+        has successfully executed its prologue hook.
+        """
+        hook_name = "prologue_logmsg"
+        hook_body = ("import pbs\n"
+                     "e = pbs.event()\n"
+                     "pbs.logjobmsg(e.job.id, 'executed prologue hook')\n")
+        attr = {'event': 'execjob_prologue', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name, attr, hook_body)
+
+        attr = {
```

Comment:
For consistency write this as-
attr = {'resources_available.ncpus': 1,
            'resources_available.mem': '2gb'}